File: sprint_1/neetcode_250/arrays_and_hashing/cpp/contains_duplicates.cpp
// https://leetcode.com/problems/contains-duplicate/

class Solution {
public:
    bool containsDuplicate(vector<int>& nums) {
        // Linear Search 
        // Time Complexity O(n^2)
        /*
        for (int i = 0; i < nums.size(); ++i) {
            for (int j = 0; j < i; ++j) {
                if (nums[i] == nums[j]) return true;
            }
        }
        return false;
        */

        // Sort and Search
        // Time Complexity O( n * log(n) )
        /* 
        sort(nums.begin(), nums.end());
        for (int i = 1; i < nums.size(); ++i) {
            if (nums[i-1] == nums[i]) return true;
        }
        return false;
        */

        // unordered_set (hash based)
        // Time Complexity O(n)
        // In practice can be slower than sorting since hashing involves a lot of backend operations
        // like hashing and collision control
        unordered_set<int> hash_set;
        for (int num: nums) {
            if (hash_set.count(num)) {
                return true;
            }
            hash_set.insert(num);
        }
        return false;
    } 
};

File: sprint_1/neetcode_250/arrays_and_hashing/python/contains_duplicates.py
# https://leetcode.com/problems/contains-duplicate/

from typing import List, Set

class Solution:
    def containsDuplicate(self, nums: List[int]) -> bool:
        # Linear Search 
        # Time Complexity O(n^2)
        """
        n: int = len(nums)

        for i in range(0, n):
            for j in range(0, i):
                if (nums[i] == nums[j]):
                    return True

        return False
        """

        # Sort and Search
        # Time Complexity O(n*log(n))
        """
        nums.sort()
        n: int = len(nums)

        for i in range(1, n):
            if (nums[i] == nums[i-1]):
                return True

        return False
        """

        # Set 
        # Time Complexity O(n)
        """
        By default in python Set and Dict are Hash based, not tree based
        Note
        This method isn't always faster than sort and search in practice, since hashing involves a lot of backend operations
        like hashing and collision control
        """

        distinct: Set[int] = set()
        for num in nums:
            if (num in distinct):
                return True
            distinct.add(num)
        return False